### PR TITLE
Feature/minsulee2/keep more api nodes

### PR DESCRIPTION
**Pull request by minsulee2**
Keep a half of whitelisted(block generation) nodes when reorg.

- Disconnect any nodes if ENABLE_JSON_RPC_API is false
- Disconnect whitelisted if they are more than majority, do not disconnect otherwise.

### CODE REVIEW

File: p2p/index.js [hunk 1]
```diff
@@ -266,6 +266,14 @@ class P2pClient {
     this.updateP2pState();
   }
 
+  pickRandomPeerAndDisconnect(addressArray) {
+    if (addressArray.length === 0) {
+      return;
+    }
+    const randomPeerAddress = _.shuffle(addressArray)[0];
+    this.closeSocketWithP2pStateUpdate(this.outbound[randomPeerAddress].socket);
```
Comment (liayoo) [inline p2p/index.js:274]: I don't think it's guaranteed that the randomPeerAddress is a key in this.outbound. Could you add a check?

Comment (minsulee2) [reply p2p/index.js:277]: Indeed! Updated! Thanks!

**Comment by minsulee2** [comment]
PTAL! All addressed!

File: p2p/index.js [hunk 2]
```diff
@@ -275,8 +286,22 @@ class P2pClient {
     const bidirectedConnections = Object.keys(this.outbound).filter(address => {
       return Object.keys(this.server.inbound).includes(address);
     });
-    const randomPeerAddress = _.shuffle(bidirectedConnections)[0];
-    this.closeSocketWithP2pStateUpdate(this.outbound[randomPeerAddress].socket);
+    if (NodeConfigs.ENABLE_JSON_RPC_API) {
+      const whitelist = this.server.node.db.getValue('/consensus/proposer_whitelist');
```
Comment (platfowner) [inline p2p/index.js:290]: 👍

File: p2p/index.js [hunk 3]
```diff
@@ -275,8 +286,22 @@ class P2pClient {
     const bidirectedConnections = Object.keys(this.outbound).filter(address => {
       return Object.keys(this.server.inbound).includes(address);
     });
-    const randomPeerAddress = _.shuffle(bidirectedConnections)[0];
-    this.closeSocketWithP2pStateUpdate(this.outbound[randomPeerAddress].socket);
+    if (NodeConfigs.ENABLE_JSON_RPC_API) {
```
Comment (platfowner) [inline p2p/index.js:289]: could you add some note that this condition is for 'API server nodes'?

File: p2p/index.js [hunk 4]
```diff
@@ -275,8 +286,22 @@ class P2pClient {
...
+      const whitelist = this.server.node.db.getValue('/consensus/proposer_whitelist');
+      const [whitelisted, notWhitelisted] =
+          _.partition(bidirectedConnections, ((address) => whitelist[address]));
+      const whitelistDisconnectThreshold = Math.round(NodeConfigs.MAX_NUM_INBOUND_CONNECTION / 2);
+      if (whitelisted.length > whitelistDisconnectThreshold) {
+        const numDisconnectionCandidates = whitelisted.length - whitelistDisconnectThreshold;
+        const randomWhiteListed = _.shuffle(whitelisted).slice(0, numDisconnectionCandidates);
+        const disconnectionCandidates = _.concat(randomWhiteListed, notWhitelisted);
```
Comment (platfowner) [inline p2p/index.js:297]: if we can define the role of the bridge nodes is to connect whitelisted nodes (proposers) and the others and interpret `whitelisted.length > whitelistDisconnectThreshold` as too less connections to non-whitelisted nodes, could we just let it try to disconnect one of the whitelisted peer connections?
this strategy can be summarized:
- too many whitelisted -> less whitelisted (i.e. disconnect) and more non-whitelisted
- too few whitelisted  -> less non-whitelisted (i.e. disconnect) and more whitelisted (in probability)

File: p2p/index.js (hunk 3 above)
Comment (platfowner) [reply p2p/index.js:289]: Also could you add a todo for introducing a separate flag as we can encounter such situations that we want to enable the json rpc api but not to assign the 'bridge node' role?

**Review by platfowner — APPROVED** [review]
LGTM with some comments.
Thanks for the quick delivery of the nice ideas and implementations!

File: p2p/index.js (hunk 3 above)
Comment (minsulee2) [reply p2p/index.js:291]: Done!

File: p2p/index.js (hunk 4 above)
Comment (minsulee2) [reply p2p/index.js:299]: Agreed. I keep it simpler for now that whitelisted sustains less than majority of nodes that is keeping 2 nodes if MAX_INBOUND equals to 5 for example. Thanks for the nice intuition!!

**Comment by minsulee2** [comment]
Thanks for the reviews! Merging now!

**Pull request merged**